File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        i = 0 
        j = len(height) - 1
        sol = 0

        while(i < j):
            if(min(height[j], height[i]) * j-i > sol):
                sol = max(min(height[j], height[i]) * (j-i), sol)
            
            diff = max(height[j], height[i]) - min(height[i], height[j])
            
            # print(i, j, sol)
            
            if(height[i] < height[j]):
                while(height[j] - height[i] >= diff):
                    i += 1
            elif(height[i] > height[j]):
                while(height[i] - height[j] >= diff):
                    j -= 1
            else:
                i += 1
                j -= 1
                
        return sol

File: Data Structures & Algorithms/eating-bananas/submission-1.java
class Solution {
    public int max(int [] nums){
        int max = Integer.MIN_VALUE;
        for(int i : nums){
            max = max > i ? max : i; 
        }
        return max;
    }
    public int eatingTime(int[] piles, int n){
        int time = 0;
        for(int i : piles){
            time = (i % n) > 0 ? time + (i/n) + 1 : time + (i/n);
        }
        return time;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int low = 1;
        int high = max(piles);
        int mid = low + (high-low) / 2;
        int res = Integer.MAX_VALUE;

        while(low <= high){
            if(eatingTime(piles, mid) > h) low = mid + 1;
            else {
                res = Math.min(mid, res);
                high = mid - 1;
            } 
            mid = low + (high - low) / 2;
        }

        return res;
    }
}
